### PR TITLE
Unify `ZQuery.foreach` implementations and cleanups

### PR DESCRIPTION
**Pull request by kyri-petrou**
The main change in this PR is that we unify the implementation for `foreach`, `foreachPar` and `foreachBatched`. This way, the `foreach` and `foreachPar` methods can take advantage of the optimizations that were done in https://github.com/zio/zio-query/pull/457.

The other somewhat big change is adding a dependency on `scala-collection-compat`. I was a bit divided about this at first, but I think it's likely for the best as I kept finding myself having to write workarounds for missing constructors / methods in Scala 2.12.

### CODE REVIEW

File: zio-query/shared/src/main/scala/zio/query/ZQuery.scala [hunk 1]
```diff
@@ -1345,26 +1336,28 @@ object ZQuery {
...
             }
             ZQuery.currentCache.getWith {
-              case cache: Cache.Default => foldPromise(cache.lookupUnsafe(request, fiberId)(Unsafe.unsafe, implicitly))
+              case cache: Cache.Default => foldPromise(cache.lookupUnsafe(request, fiberId)(Unsafe.unsafe))
               case cache                => cache.lookup(request).flatMap(foldPromise)
             }
-          } else
-            Promise.makeAs[E, B](fiberId).map { promise =>
+          } else {
+            ZIO.succeed {
```
Comment (ghostdogpr) [inline zio-query/shared/src/main/scala/zio/query/ZQuery.scala:1353]: Out of curiosity, why this change? Isn't `Promise.makeAs` doing the same thing?

**Review by ghostdogpr — DISMISSED** [review]

File: zio-query/shared/src/main/scala/zio/query/ZQuery.scala (hunk 1 above)
Comment (kyri-petrou) [reply zio-query/shared/src/main/scala/zio/query/ZQuery.scala:1353]: Because when using `Promise.makeAs` we have to use `.map` afterwards (which is an extra flatMap). Since this is run for every `ZQuery.fromRequest`, I think it's worth to use the `unsafe` method in order to reduce the overall number of flatMaps in the query

Comment (ghostdogpr) [reply zio-query/shared/src/main/scala/zio/query/ZQuery.scala:1353]: Ah right, I missed that
